src/Servers.jl

```diff
@@ -130,278 +256,206 @@ function check_rate_limit(tcp;
+ - `reuseaddr::Bool=false`: whether multiple servers should be allowed to listen on the same port
  - `tcpref::Ref{Base.IOServer}`, this reference is set to the underlying
                                  `IOServer`. e.g. to allow closing the server.
+ - `connectioncounter::Ref{Int}`: a `Ref{Int}` that can be used to track the # of currently open (i.e
+        currently being handled) connections for a server
+ - `ratelimit`: a `Rational{Int}` of the form `5//1` indicating how many `messages//second`
+        should be allowed per client IP address; requests exceeding the rate limit will be auto-closed
+ - `stream::Bool=false`: whether the handler should operate on an `HTTP.Stream` to read & write directly
+ - `verbose::Bool=false`: whether simple logging should print to stdout for connections handled
 
 e.g.
 ```
-    HTTP.listen() do http::HTTP.Stream
+    HTTP.listen(; stream=true) do http::HTTP.Stream
         @show http.message
         @show HTTP.header(http, "Content-Type")
         while !eof(http)
             println("body data: ", String(readavailable(http)))
         end
-        HTTP.setstatus(http, 404)
-        HTTP.setheader(http, "Foo-Header" => "bar")
+        setstatus(http, 404)
+        setheader(http, "Foo-Header" => "bar")
         startwrite(http)
         write(http, "response body")
         write(http, "more response body")
+        return
     end
 
     HTTP.listen() do request::HTTP.Request
         @show HTTP.header(request, "Content-Type")
-        @show HTTP.payload(request)
+        @show HTTP.payload(request, String)
         return HTTP.Response(404)
     end
 ```
 """
-listen(f, host::IPAddr=Sockets.localhost, port::Integer=8081; kw...) = listen(f, Sockets.InetAddr(host, port); kw...)
-listen(f, host::AbstractString, port::Integer; kw...) = listen(f, parse(IPAddr, host), port; kw...)
-listen(f, host::AbstractString; kw...) = listen(f, string(host); kw...)
-
-function listen(f::Function,
-                host::Union{Sockets.InetAddr, String},
-                ;
-                ssl::Bool=false,
-                require_ssl_verification::Bool=true,
-                sslconfig::Union{SSLConfig, Nothing}=nothing,
-                pipeline_limit::Int=ConnectionPool.default_pipeline_limit,
-                tcpisvalid::Function=(tcp; kw...)->true,
-                tcpref::Ref=Ref{Base.IOServer}(),
-                reuseaddr::Bool=false,
-                connectioncounter::Base.RefValue{Int}=Ref(0),
-                kw...)
-
-    if ssl && sslconfig === nothing
-        sslconfig = SSLConfig(require_ssl_verification)
-    end
-
-    @info "Listening on: $host"
-    if isassigned(tcpref)
+function listen end
+
+getinet(host::String, port::Integer) = Sockets.InetAddr(parse(IPAddr, host), port)
+getinet(host::IPAddr, port::Integer) = Sockets.InetAddr(host, port)
+
+listen(f::Base.Callable, args...; stream::Bool=false, kw...) =
```

Comment:
Does it make sense to keep this as `Base.Callable` (since everything in Julia is callable nowadays)?